### ISSUE TITLE
issue in highlighting fortran code: kind of intrinsic data type

### ISSUE THREAD

**Issue by kookma** [highlightjs/highlight.js#1723]
In modern fortran you can set the kind of intrinsic data type. It means you can define additional intrinsic data type. When you do so, for literal constants you have to use the kind name attached to literal consonant using under score.
## Example

```fortran
PROGRAM test_initial
INTEGER, PARAMETER :: DBL = SELECTED_REAL_KIND(p=13)
REAL(KIND=DBL) :: a1 = 6.666666666666666
REAL(KIND=DBL) :: a2 = 6.666666666666666_DBL
WRITE (*,*) a1, a2
END PROGRAM test_initial
```
## Highlight.js issue
When using highlight.js to highlight a Fortran code contains such named constant, highlight.js fails to do the job! It highlights the remaining code (from the  position of named constant) in a wrong way!  

## Note

- The above code in this post is highlighted by github, it works correctly! See the name constant is in blue color.
- other intrinsic data can have other kind. But in all cases in named constants, the kind is specified by an underscore to the constant (e.g  `12_ps`)

**Comment by kookma** [comment]
**Solution**

This is the solution to the above question.
The "number" class in `fortran.js` file shall be corrected as bellow:

```
{
        className: 'number',
        begin: '(?=\\b|\\+|\\-|\\.)[+-]?(?:\\.|\\d+\\.?)\\d*([de][+-]?\\d+)?(_[a-z\\d]+)?\\b',
        relevance: 0
      }

```

**Comment by kookma** [comment]
Is there any chance to be merged in the new revision of main release?

**Comment by egor-rogov** [comment]
@kookma Could you make a PR out of this (plus some tests)?

**Comment by egor-rogov** [comment]
Yes, first fork the repo and commit changes to it. Then in your repo on "Pull requests" tab find "New pull request" button.

**Comment by kookma** [comment]
@egor-rogov 
Please have a look at: https://github.com/highlightjs/highlight.js/pull/1934

**Comment by yyyc514** [comment]
See also #1566

**Comment by kookma** [comment]
@yyyc514 
I check the new fix on https://jsfiddle.net/ajoshguy/nagkqytv/11/
by trying the below code:
```
<pre>
<code class="fortran">
program main
! This is a test
real :: x, y
x=1.23_wp
y=0._wer
z=1.25e9
end program main
</code>
</pre>
```
And it seems you solution from #1566 works!

So, I think you can to close this issue.

**Comment by yyyc514** [comment]
Well, our desire is the highlight the WHOLE number, including the "kind", that's why this is still open.  What should be fixed now if us "losing" the first digit, which was not good at all.

**Comment by kookma** [comment]
@yyyc514 

What I checked at   https://jsfiddle.net/ajoshguy/nagkqytv/11/ is
<pre>
<code class="fortran">
program main
! This is a test
real :: x, y
x=1.23_wp
y=0._wer
z=1.25e9
end program main
</code>
</pre>

and it shows the whole number including the first number and the kind highlighted!
I mean number like 12.34_wp

I am afraid I don't understand what you explained.

**Comment by yyyc514** [comment]
![Screen Shot 2019-10-24 at 3 37 15 PM](https://user-images.githubusercontent.com/6473/67519143-33f76000-f674-11e9-9576-45f97b30c77b.png)

Not sure what you are seeing, but this is not correct yet.  And it can't be since the regex hasn't been updated. Its' possible you're seeing some other language highlighting the code instead.

**Comment by kookma** [comment]
I see it as below, so it works at https://jsfiddle.net/ajoshguy/nagkqytv/11/
![hilight-fortran](https://user-images.githubusercontent.com/830394/67519531-e3075c00-f6b3-11e9-9893-0cdf2cd64249.png)

**Comment by yyyc514** [comment]
Not Fortran. :-)  Did you tell it to load fortran in the init config?

**Comment by kookma** [comment]
Where is that? I just used `<code class="fortran">`
I assume it loads all language brushes!

**Comment by yyyc514** [comment]
No it loads only the common set. It's at the top of the file.

**Comment by kookma** [comment]
@yyyc514  okay, I changed the first like to `var language = "fortran";` and now I see what you posted!

I my revised version as I explained above I use

```
{
        className: 'number',
        begin: '(?=\\b|\\+|\\-|\\.)[+-]?(?:\\.|\\d+\\.?)\\d*([de][+-]?\\d+)?(_[a-z\\d]+)?\\b',
        relevance: 0
      }
```
and it works fine. Have a look at

https://kookma.github.io/TW-Scripts/#Syntax%20Highlighting:%5B%5BSyntax%20Highlighting%5D%5D%20%24%3A%2Fplugins%2Ftiddlywiki%2Fhighlight%20%24%3A%2Fplugins%2Ftiddlywiki%2Fhighlight%2Fhighlight.js

Here I use highlight.js as plugin! Scroll down to see the modified highlight.js

**Comment by kookma** [comment]
You can edit online the content and check Fortran code of your own!
Edit change and then push the save button!

**Comment by yyyc514** [comment]
Yeah, we've been debating on the correct/best rule.  That might wait till .12 though, depends how much time I have between now and release.

My priority was first fixing the data loss.

**Comment by kookma** [comment]
@yyyc514 
I have checked 9.18.1 the issue https://github.com/highlightjs/highlight.js/issues/1723#issuecomment-546070933

still there!

**Comment by yyyc514** [comment]
The full fix should be in 10.0.  Pretty sure we closed this one down.  Please test on the 10.0 beta.

**Comment by kookma** [comment]
many thanks Josh!
